### ISSUE TITLE
[java] JUnitTestsShouldIncludeAssert false positive with inherited @Rule field

### ISSUE THREAD

**Issue by blerner** [pmd/pmd#1422]
**Affects PMD Version:** 6.9.0

**Rule:** JUnitTestsShouldIncludeAssertRule

**Description:** Thanks to bug #631, this rule properly looks for `@Rule ExceptedException` fields, and correctly does not complain about tests that use them.  However, it only checks for such rules in the current class.  One scenario that's come up in my students' code is writing an abstract test class with a bunch of concrete tests and an abstract factory method, and then several concrete test classes that instantiate the factory with one of several implementations.  Additionally, they write some concrete tests in the concrete classes.  Now, the `@Rule` isn't a declared field in the same class as the `@Test`s that use it, so this rule incorrectly fires again.

I think the solution is extending [here](https://github.com/pmd/pmd/blob/439800495321e7ab4a023938a619c1815ef4fddf/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/JUnitTestsShouldIncludeAssertRule.java#L79) to walk up the inheritance tree, so far as it's known (and here, it should be known), and check for inherited `@Rule` fields...

**Code Sample demonstrating the issue:**

```java
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

abstract class Shape {
  double size;

  Shape(double size) { 
    if (size < 0) {
      throw new IllegalArgumentException("negative size");
    }
    this.size = size; 
  }

  public abstract double area();
}

class Square extends Shape {
  Square(double size) {
    super(size);
  }
  
  public double area() {
    return size * size;
  }
}

class Weird extends Shape {
  Weird(double size) {
    super(size);
  }
  
  public double area() {
    throw new RuntimeException("just 'cause");
  }
}

public abstract class ShapeTester {
  protected abstract Shape factory(double size);

  @Rule
  public ExpectedException e = ExpectedException.none();

  @Test 
  public void notAProblem() {
    e.expect(IllegalArgumentException.class);
    factory(-5);
  }

  public static final class SquareTester extends ShapeTester {
    protected Shape factory(double size) {
      return new Square(size);
    }
  }

  public static final class WeirdTester extends ShapeTester {
    protected Shape factory(double size) {
      return new Weird(size);
    }

    @Test
    public void thisOneTriggersRule() {
      e.expect(RuntimeException.class);
      e.expectMessage("just 'cause");
      factory(42).area();
    }
  }
}
```

**Running PMD through:** programmatically

**Comment by adangel** [comment]
Fixed via #2899 for PMD 7.

**Comment by Sandm-HY** [comment]
Hello, so is this issue resolved now?

**Comment by jsotuyod** [comment]
Yes, PMD 7 RCs include this fix.
